Notify about recently updated packages

DragoonAethis opened the pull request:
This PR adds an extra green text line with "X minutes ago" to recently upgraded packages, next to the update date. This is to notify people that a given version might not yet be synchronized on various mirrors. Looks like this:

![image](https://user-images.githubusercontent.com/1546665/101246839-ffc6ed80-3715-11eb-842c-0ce32e8df5c8.png)

FFY00 reviewed (approved):
Thanks, people keep flagging discord :disappointed:

DragoonAethis commented:
And whine about it on IRC :P Maybe it could be orange/red, or get highlighted on "flag package" link hover so that it's easier to notice right away?

jelly commented on main/models.py:
```diff
@@ -133,6 +135,15 @@ def get_full_url(self, proto='https'):
+
+    @property
+    def is_recent(self):
+        return self.updated_mins_ago <= 90
```
Is this an educated guess or well informed?

DragoonAethis replied:
I took the mirror sync delays from the [mirror status](https://www.archlinux.org/mirrors/status/), removed all mirrors with more than 6 hours of delay (around 100 of them), then averaged the rest (around 600) - that gives you 1h26min, so I just rounded it up to 90 minutes. This notifies about delays on about 400 mirrors.

Averaging mirrors below 10 hours gives 1h44mins (covers 420 mirrors), below 24 hours gives 3h06mins (covers 500 mirrors) and all mirrors average to about 4 hours (covers 550 mirrors). The longer this period is, the more annoying this notification ends up being for maintainers though, so I took the first average as a good starting point. If people keep complaining too much, it can be always bumped up.

DragoonAethis commented:
Updated to diff time on timedelta.total_seconds() and not timedelta.seconds, since that makes the message display every day (oops). Also added an underline to the "X mins ago" text so that it's a bit more obvious this has an alt text.

jelly commented:
Sorry for the delay, found a small bug:
```
2020-12-30 21:09 UTC (0.0 minutes ago)
--

DragoonAethis commented:
TIL floats with nothing after the decimal point will be printed as whole numbers by the template engine, but not if it's 0.0... I'll add a cast there.

Date of archweb import vs repo update import is a concern, but in this case the database doesn't contain any more concrete info about when a package was actually published in any repositories. One could check for the build date, but that breaks too eg with package rebuilds (where you might want to wait for tens of packages to be built before pushing). I'm not sure if there's any relatively clean way to do this without adding an extra repo_update DateTimeField, plus some extra logic for repo update scripts to set that in the archweb database. That sounds nastier than just being late with the archweb notification.

jelly commented:
> TIL floats with nothing after the decimal point will be printed as whole numbers by the template engine, but not if it's 0.0... I'll add a cast there.
> 
> Date of archweb import vs repo update import is a concern, but in this case the database doesn't contain any more concrete info about when a package was actually published in any repositories. One could check for the build date, but that breaks too eg with package rebuilds (where you might want to wait for tens of packages to be built before pushing). I'm not sure if there's any relatively clean way to do this without adding an extra repo_update DateTimeField, plus some extra logic for repo update scripts to set that in the archweb database. That sounds nastier than just being late with the archweb notification.

Yup, we'll take it as is and I'll try to remember that :+1:

DragoonAethis commented:
Fixed, added a comment for future reference.

DragoonAethis commented:
While we're around this topic, maybe I could also add something like a red bar on the flagging page to be extra sure? Something like this:

![image](https://user-images.githubusercontent.com/1546665/103382737-5bf01800-4af0-11eb-9506-60dbe8f3bb60.png)
